cl/merkle_tree/merkle_tree.go

```diff
@@ -0,0 +1,249 @@
+package merkle_tree
+
+import (
+	"bytes"
+
+	libcommon "github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon-lib/common/length"
+)
+
+const OptimalMaxTreeCacheDepth = 12
+
+type MerkleTree struct {
+	computeLeaf func(idx int, out []byte)
+	layers      [][]byte // Flat hash-layers
+	leavesCount int
+
+	hashBuf [64]byte // buffer to store the input for hash(hash1, hash2)
+	limit   *uint64  // Optional limit for the number of leaves (this will enable limit-oriented hashing)
+}
+
+// Layout of the layers:
+
+// 0-n: intermediate layers
+// Root is not stored in the layers, root is recomputed on demand
+// The first layer is not the leaf layer, but the first intermediate layer, the leaf layer is not stored in the layers.
+
+// Initialize initializes the Merkle tree with the given number of leaves and the maximum depth of the tree cache.
+func (m *MerkleTree) Initialize(leavesCount, maxTreeCacheDepth int, computeLeaf func(idx int, out []byte), limitOptional *uint64) {
+	m.computeLeaf = computeLeaf
+	m.layers = make([][]byte, maxTreeCacheDepth)
+	m.leavesCount = leavesCount
+	firstLayerSize := ((leavesCount + 1) / 2) * length.Hash
+	capacity := (firstLayerSize / 2) * 3
+	m.layers[0] = make([]byte, firstLayerSize, capacity)
+	if limitOptional != nil {
+		m.limit = new(uint64)
+		*m.limit = *limitOptional
+	}
+}
+
+// MarkLeafAsDirty resets the leaf at the given index, so that it will be recomputed on the next call to ComputeRoot.
+func (m *MerkleTree) MarkLeafAsDirty(idx int) {
+	for i := 0; i < len(m.layers); i++ {
+		currDivisor := 1 << (i + 1) // i+1 because the first layer is not the leaf layer
+		layerSize := (m.leavesCount + (currDivisor - 1)) / currDivisor
+		if layerSize == 0 {
+			break
+		}
+		if m.layers[i] == nil {
+			capacity := (layerSize / 2) * 3
+			if capacity == 0 {
+				capacity = 1024
+			}
+			m.layers[i] = make([]byte, layerSize, capacity)
+		}
+		copy(m.layers[i][(idx/currDivisor)*length.Hash:], ZeroHashes[0][:])
+		if layerSize == 1 {
+			break
+		}
+	}
+}
+
+func (m *MerkleTree) AppendLeaf() {
+	/*
+		Step 1: Append a new dirty leaf
+		Step 2: Extend each layer with the new leaf when needed (1.5x extension)
+	*/
+	for i := 0; i < len(m.layers); i++ {
```

Comment:
I am not sure what will happen if:
1. initialize with leaf size 2
2. AppendLeaf 3 times with non-empty leaf (i know which is fetched in m.computeLeaf(...))
3. then compute root

so in this case, the root layer turns from Layer0 to Layer2,  but this loop seems not extend new layers?
or, extend new layers while computing root in Line154? https://github.com/erigontech/erigon/pull/11096/files?diff=unified&w=0#diff-5c053bedbcbe2473e56d2a0b317ba9e2f4d6f169faaf83e6b5bdd463cc2b7bceR154 
ignore me if this case can be properly handled

Reply:
It is handled:


## Initial State

```
        root
leaf_1  leaf_2
```

## Adding Two Leafs

When adding a new leaf node, the following thing occur:

```
                               root (extended)
                              /                 \
            intermediate_node_1       intermediate_node_2 (newly added)
           /        \                   /           \
      leaf_1     leaf_2            leaf_3     empty_leaf (newly added)
```

## Steps Explained

- **Initial State**: 
    - The tree starts with a `root` and two leaves: `leaf_1` and `leaf_2`.

- **First Extension**:
    - When extending the tree, an intermediate layer is appended to the root.
    - `intermediate_node_1` and `intermediate_node_2` are introduced.
    - `leaf_3` and an `empty_leaf` are added as children of `intermediate_node_2`.

The structure now has an extended root with intermediate nodes and new leaf nodes.